$OpenBSD: patch-avahi-core_socket_c,v 1.1 2010/07/07 18:57:38 ajacoutot Exp $

Fix for CVE-2010-2244: 2b2844b10d7b7e5c97f9c667d664d9418bb7769a
    zero size is reported for corrupt packets. recvmsg() later could
    nevertheless get data from a good packet that followed the bad one.
    So get out early to avoid hitting an assertion.

--- avahi-core/socket.c.orig	Wed Jun 18 01:13:44 2008
+++ avahi-core/socket.c	Wed Jul  7 20:53:39 2010
@@ -652,6 +652,10 @@ AvahiDnsPacket *avahi_recv_dns_packet_ipv4(
         goto fail;
     }
 
+    /* For corrupt packets FIONREAD returns zero size (See rhbz #607297) */
+    if (!ms)
+        goto fail;
+
     p = avahi_dns_packet_new(ms + AVAHI_DNS_PACKET_EXTRA_SIZE);
 
     io.iov_base = AVAHI_DNS_PACKET_DATA(p);
@@ -804,6 +808,10 @@ AvahiDnsPacket *avahi_recv_dns_packet_ipv6(
         avahi_log_warn("FIONREAD returned negative value.");
         goto fail;
     }
+
+    /* For corrupt packets FIONREAD returns zero size (See rhbz #607297) */
+    if (!ms)
+        goto fail;
 
     p = avahi_dns_packet_new(ms + AVAHI_DNS_PACKET_EXTRA_SIZE);
 
